policy_agent.py

```diff
@@ -216,32 +194,54 @@ def _process_update(self, event_type, resource_type, resource):
             _log.debug("Calling handler: %s", handler)
             try:
                 handler(key, resource)
+                _log.info("%s %s: %s", event_type, resource_type, key)
             except KeyError:
                 _log.exception("Invalid %s: %s", resource_type,
                                json.dumps(resource, indent=2))
 
-    def _watch_api(self, path, resource_version=None):
+    def _manage_resource(self, resource_type):
         """
-        Work loop for the watch thread.
+        Routine for a worker thread.  Syncs with API for the given resource
+        and starts a watch.  If an error occurs within the watch, will re-sync
+        with the API and re-start the watch.
         """
-        _log.info("Starting watch on path: %s", path)
         while True:
-            # Attempt to stream API resources.
             try:
-                response = self._api_get(path,
-                                         stream=True,
-                                         resource_version=resource_version)
-                _log.debug("Watch response for %s: %s", path, response)
+                # Sync existing resources for this type.
+                resource_version = self._sync_resources(resource_type)
+
+                # Start a watch from the latest resource_version.
+                self._watch_resource(resource_type, resource_version)
             except requests.ConnectionError:
-                _log.exception("Error querying path: %s", path)
-                time.sleep(10)
-                continue
+                _log.exception("Connection error querying: %s", resource_type)
+            except requests.HTTPError:
+                _log.exception("HTTP error querying: %s", resource_type)
+            except KubernetesApiError:
+                _log.exception("Kubernetes API error managing %s",
```

Comment:
This might be too strong a log at scale.  Let's see how it turns out, but it might be the case that we end up degrading from watching into regular resyncs (Felix does this).